hyper/http11/response.py

```diff
@@ -48,7 +49,8 @@ def __init__(self, code, reason, headers, sock, connection=None):
         # bother checking for content-length, we just keep reading until
         # we no longer can.
         self._expect_close = False
-        if b'close' in self.headers.get(b'connection', []):
+        if (b'close' in self.headers.get(b'connection', []) or
+                b'close' in self.headers.get(b'proxy-connection', [])):
             self._expect_close = True
```

Comment:
> the spec also requires that in the absence of all three headers (no Connection: close, no Transfer-Encoding: chunked, and no Content-Length), the response must be treated as if Connection: close were present.

I guess this quote confirms what you say:

[RFC 7230, HTTP/1.1 Message Syntax and Routing, section 3.3.3. Message Body Length](https://tools.ietf.org/html/rfc7230#section-3.3.3)
> The length of a message body is determined by one of the following
> (in order of precedence):
> 
> [... cases when either `content-length` or `transfer-encoding: chunked` are set ...]
> 
> 7.  Otherwise, this is a response message without a declared message
>    body length, so the message body length is determined by the
>    number of octets received prior to the server closing the
>    connection.
> 
> Since there is no way to distinguish a successfully completed,
> close-delimited message from a partially received message interrupted
> by network failure, a server SHOULD generate encoding or
> length-delimited messages whenever possible.  The close-delimiting
> feature exists primarily for backwards compatibility with HTTP/1.0.

Here are some extracts that doesn't say the opposite, but it's implicitly meant that in the absence of `Connection: close` header for HTTP/1.1 response a client should persist connection.

[RFC 7230, HTTP/1.1 Message Syntax and Routing, section 6.3. Persistence](https://tools.ietf.org/html/rfc7230#section-6.3):

> o  If the "close" connection option is present, the connection will
>   not persist after the current response; else,
> 
> o  If the received protocol is HTTP/1.1 (or later), the connection
>   will persist after the current response; else,
> 
> o  If the received protocol is HTTP/1.0, the "keep-alive" connection
>   option is present, the recipient is not a proxy, and the recipient
>   wishes to honor the HTTP/1.0 "keep-alive" mechanism, the
>   connection will persist after the current response; otherwise,
> 
> o  The connection will close after the current response.


[RFC 2068, HTTP/1.1, section 8.1.2.1 Negotiation](https://tools.ietf.org/html/rfc2068#section-8.1.2.1):

> An HTTP/1.1 client MAY expect a connection to remain open, but would
> decide to keep it open based on whether the response from a server
> contains a Connection header with the connection-token close. [...]
> 
> If either the client or the server sends the close token in the
> Connection header, that request becomes the last one for the
> connection.
> 
> Clients and servers SHOULD NOT assume that a persistent connection is
> maintained for HTTP versions less than 1.1 unless it is explicitly
> signaled. See section 19.7.1 for more information on backwards
> compatibility with HTTP/1.0 clients.

In sum, we have that in the absence of `content-length` and `transfer-encoding: chunked` headers connection should be closed, but this intent must be explicitly defined with the `connection: close` header. Thus I assume that it's not a valid HTTP/1.1 response and I guess that the assertion we discuss should be triggered for HTTP/1.1 responses, but should not for HTTP/1.0.